From de5afd3a45c3bfa49e595f6b49c57fa8eef376bd Mon Sep 17 00:00:00 2001
From: Gustavo Solaira <gustavos@codeaurora.org>
Date: Fri, 24 Jul 2015 18:34:51 -0700
Subject: [PATCH] Change Eagle to correct PID 28

---
 arch/arm/mach-msm/socinfo.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/arch/arm/mach-msm/socinfo.c b/arch/arm/mach-msm/socinfo.c
index 13b2d08..864ab16 100644
--- a/arch/arm/mach-msm/socinfo.c
+++ b/arch/arm/mach-msm/socinfo.c
@@ -53,13 +53,11 @@ enum {
 	HW_PLATFORM_SVLTE_SURF	= 5,
 	HW_PLATFORM_MTP  = 8,
 	HW_PLATFORM_LIQUID  = 9,
-	/* Dragonboard platform id is assigned as 10 in CDT */
-    /* Using Same ID for Eagle for now. TBD */
 	HW_PLATFORM_DRAGON	= 10,
 	HW_PLATFORM_QRD	= 11,
 	HW_PLATFORM_HRD	= 13,
 	HW_PLATFORM_DTV	= 14,
-	HW_PLATFORM_NPC_EAGLE	= 23,
+	HW_PLATFORM_EAGLE = 28,
 	HW_PLATFORM_INVALID
 };
 
@@ -76,7 +74,7 @@ const char *hw_platform[] = {
 	[HW_PLATFORM_QRD] = "QRD",
 	[HW_PLATFORM_HRD] = "HRD",
 	[HW_PLATFORM_DTV] = "DTV",
-	[HW_PLATFORM_NPC_EAGLE] = "NPC Eagle",
+	[HW_PLATFORM_EAGLE] = "EAGLE",
 };
 
 enum {
-- 
2.4.2

