diff --git a/media/mtp/MtpMain.cpp b/media/mtp/MtpMain.cpp
index a401379..ceb6f51 100644
--- a/media/mtp/MtpMain.cpp
+++ b/media/mtp/MtpMain.cpp
@@ -1040,7 +1040,7 @@ MtpProperty* MyMtpDatabase::getDevicePropertyDesc(MtpDeviceProperty property)
         case MTP_DEVICE_PROPERTY_SYNCHRONIZATION_PARTNER:
         case MTP_DEVICE_PROPERTY_DEVICE_FRIENDLY_NAME:
             result = new MtpProperty(property, MTP_TYPE_STR, true);
-            result->setCurrentValue(str.string());
+            result->setCurrentValue(reinterpret_cast<const uint16_t*>(str.string()));
             break;
         default:
             break;
