diff --git a/scripts/gcc-wrapper.py b/scripts/gcc-wrapper.py
index 27e76a7..c5409dc 100755
--- a/scripts/gcc-wrapper.py
+++ b/scripts/gcc-wrapper.py
@@ -43,6 +43,16 @@ allowed_warnings = set([
     "alignment.c:327",
     "mmu.c:602",
     "return_address.c:62",
+    "transport.c:653",
+    "swab.h:49",
+    "xt_socket.c:307",
+    "xt_socket.c:161",
+    "inet_hashtables.h:356",
+    "epm_adc.c:520",
+    "epm_adc.c:644",
+    "pm8xxx-ccadc.c:825",
+    "xc4000.c:1049",
+    "xc4000.c:1063",
  ])
 
 # Capture the name of the object file, can find it.
