From 80523ab4eeebc9c8356b7b1f9dd036a1370c9549 Mon Sep 17 00:00:00 2001
From: Kaustubh Gondkar <kgondkar@codeaurora.org>
Date: Tue, 12 Nov 2013 17:52:10 -0800
Subject: [PATCH] SOM Disable Audio devices

---
 arch/arm/mach-msm/board-8064.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/arch/arm/mach-msm/board-8064.c b/arch/arm/mach-msm/board-8064.c
index 83f24e9..b76ab02 100644
--- a/arch/arm/mach-msm/board-8064.c
+++ b/arch/arm/mach-msm/board-8064.c
@@ -3364,9 +3364,11 @@ static void __init apq8064_common_init(void)
 			platform_device_register(&mdm_8064_device);
 		}
 	}
-	platform_device_register(&apq8064_slim_ctrl);
-	slim_register_board_info(apq8064_slim_devices,
-		ARRAY_SIZE(apq8064_slim_devices));
+	if(!machine_is_apq8064_som()) {
+		platform_device_register(&apq8064_slim_ctrl);
+		slim_register_board_info(apq8064_slim_devices,
+			ARRAY_SIZE(apq8064_slim_devices));
+	}
 	if (!PLATFORM_IS_MPQ8064()) {
 		apq8064_init_dsps();
 		platform_device_register(&msm_8960_riva);
-- 
1.8.2.1

