From 0c101056ac8c6f79f5f987842bc1b649648fcfca Mon Sep 17 00:00:00 2001
From: Kaustubh Gondkar <kgondkar@codeaurora.org>
Date: Mon, 25 Nov 2013 17:04:08 -0800
Subject: [PATCH] SOM I2C host controller configuration

---
 arch/arm/mach-msm/board-8064.c | 45 ++++++++++++++++++++++++++++--------------
 1 file changed, 30 insertions(+), 15 deletions(-)

diff --git a/arch/arm/mach-msm/board-8064.c b/arch/arm/mach-msm/board-8064.c
index b76ab02..a06486b 100644
--- a/arch/arm/mach-msm/board-8064.c
+++ b/arch/arm/mach-msm/board-8064.c
@@ -2723,6 +2723,13 @@ static void __init apq8064_i2c_init(void)
 {
 	void __iomem *gsbi_mem;
 
+	if(machine_is_apq8064_som()) {
+		/* SOM TBD: gsbi3 and gsbi4 pdata needs to be updated */
+		apq8064_device_qup_i2c_gsbi3.dev.platform_data =
+					&apq8064_i2c_qup_gsbi3_pdata;
+		apq8064_device_qup_i2c_gsbi4.dev.platform_data =
+					&apq8064_i2c_qup_gsbi4_pdata;
+	} else {
 	apq8064_device_qup_i2c_gsbi1.dev.platform_data =
 					&apq8064_i2c_qup_gsbi1_pdata;
 	gsbi_mem = ioremap_nocache(MSM_GSBI1_PHYS, 4);
@@ -2744,6 +2751,7 @@ static void __init apq8064_i2c_init(void)
 	}
 	mpq8064_device_qup_i2c_gsbi5.dev.platform_data =
 					&mpq8064_i2c_qup_gsbi5_pdata;
+	} 
 }
 
 #if defined(CONFIG_KS8851) || defined(CONFIG_KS8851_MODULE)
@@ -3023,6 +3031,7 @@ static void __init apq8064_init_dsps(void)
 #define I2C_MPQ_CDP	BIT(5)
 #define I2C_MPQ_HRD	BIT(6)
 #define I2C_MPQ_DTV	BIT(7)
+#define I2C_APQ_SOM	BIT(7) /* For APQ SOM I2C device */
 
 struct i2c_registry {
 	u8                     machs;
@@ -3062,6 +3071,15 @@ static struct i2c_registry apq8064_i2c_devices[] __initdata = {
 		cs8427_device_info,
 		ARRAY_SIZE(cs8427_device_info),
 	},
+#ifdef CONFIG_NFC_NXP_PN512_I2C
+/* SOM TBD: NFC PN512 support */
+	{
+		I2C_APQ_SOM,
+		APQ_8064_GSBI3_QUP_I2C_BUS_ID,
+		nxp_pn544_device_info,
+		ARRAY_SIZE(nxp_pn512_device_info),
+	},
+#endif
 };
 
 #define SX150X_EXP1_INT_N	PM8921_MPP_IRQ(PM8921_IRQ_BASE, 9)
@@ -3144,8 +3162,6 @@ static void __init register_i2c_devices(void)
 	int i;
 
 #ifdef CONFIG_MSM_CAMERA
-/* SOM: Disable camera for bringup */
-#ifndef CONFIG_MACH_APQ8064_SOM
 	struct i2c_registry apq8064_camera_i2c_devices = {
 		I2C_SURF | I2C_FFA | I2C_LIQUID | I2C_RUMI,
 		APQ_8064_GSBI4_QUP_I2C_BUS_ID,
@@ -3153,7 +3169,6 @@ static void __init register_i2c_devices(void)
 		apq8064_camera_board_info.num_i2c_board_info,
 	};
 #endif
-#endif
 	/* Build the matching 'supported_machs' bitmask */
 	if (machine_is_apq8064_cdp())
 		mach_mask = I2C_SURF;
@@ -3163,6 +3178,8 @@ static void __init register_i2c_devices(void)
 		mach_mask = I2C_LIQUID;
 	else if (PLATFORM_IS_MPQ8064())
 		mach_mask = I2C_MPQ_CDP;
+	else if (machine_is_apq8064_som())
+		mach_mask = I2C_APQ_SOM;
 	else
 		pr_err("unmatched machine ID in register_i2c_devices\n");
 
@@ -3174,13 +3191,14 @@ static void __init register_i2c_devices(void)
 						apq8064_i2c_devices[i].len);
 	}
 #ifdef CONFIG_MSM_CAMERA
-	/* SOM: Disable camera temp */
-#ifndef CONFIG_MACH_APQ8064_SOM
-	if (apq8064_camera_i2c_devices.machs & mach_mask)
-		i2c_register_board_info(apq8064_camera_i2c_devices.bus,
-			apq8064_camera_i2c_devices.info,
-			apq8064_camera_i2c_devices.len);
-#endif
+	/* SOM: no camera i2c device */
+	if (!machine_is_apq8064_som()) {
+		if (apq8064_camera_i2c_devices.machs & mach_mask) {
+			i2c_register_board_info(apq8064_camera_i2c_devices.bus,
+				apq8064_camera_i2c_devices.info,
+				apq8064_camera_i2c_devices.len);
+		}
+	}
 #endif
 
 	for (i = 0; i < ARRAY_SIZE(mpq8064_i2c_devices); ++i) {
@@ -3274,11 +3292,8 @@ static void __init apq8064_common_init(void)
 		pr_err("Failed to initialize XO votes\n");
 	msm_clock_init(&apq8064_clock_init_data);
 	apq8064_init_gpiomux();
-	/* SOM: Disable I2C devices */
-	if(!machine_is_apq8064_som()) { 
-		apq8064_i2c_init();
-		register_i2c_devices();
-	}
+	apq8064_i2c_init();
+	register_i2c_devices();
 
 	apq8064_device_qup_spi_gsbi5.dev.platform_data =
 						&apq8064_qup_spi_gsbi5_pdata;
-- 
1.8.2.1

