diff --git a/arch/arm/mach-msm/board-8064-camera.c b/arch/arm/mach-msm/board-8064-camera.c
index 6f1f0b8..42cdbf7 100644
--- a/arch/arm/mach-msm/board-8064-camera.c
+++ b/arch/arm/mach-msm/board-8064-camera.c
@@ -441,8 +441,8 @@ static struct camera_vreg_t apq_8064_cam_vreg[] = {
 	{"cam_vaf", REG_LDO, 2800000, 2850000, 300000},
 };
 
-#define CAML_RSTN PM8921_GPIO_PM_TO_SYS(28)
-#define CAMR_RSTN 34
+#define  CAML_RSTN PM8921_GPIO_PM_TO_SYS(28)
+#define CAMR_RSTN 34
 
 static struct gpio apq8064_common_cam_gpio[] = {
 #if 0
@@ -790,6 +790,7 @@ void __init apq8064_init_cam(void)
 
 #ifdef CONFIG_I2C
 static struct i2c_board_info apq8064_camera_i2c_boardinfo[] = {
+#if 0
 	{
 	I2C_BOARD_INFO("imx074", 0x1A),
 	.platform_data = &msm_camera_sensor_imx074_data,
@@ -806,12 +807,14 @@ static struct i2c_board_info apq8064_camera_i2c_boardinfo[] = {
 	I2C_BOARD_INFO("ov2720", 0x6C),
 	.platform_data = &msm_camera_sensor_ov2720_data,
 	},
+#endif
 #ifdef CONFIG_OV5640
 	{
 	I2C_BOARD_INFO("ov5640", 0x3c),
 	.platform_data = &msm_camera_sensor_ov5640_data,
 	},
 #endif
+#if 0
 	{
 	I2C_BOARD_INFO("sc628a", 0x6E),
 	},
@@ -823,6 +826,7 @@ static struct i2c_board_info apq8064_camera_i2c_boardinfo[] = {
 	I2C_BOARD_INFO("s5k3l1yx", 0x20),
 	.platform_data = &msm_camera_sensor_s5k3l1yx_data,
 	},
+#endif
 };
 
 struct msm_camera_board_info apq8064_camera_board_info = {
