diff --git a/arch/arm/mach-msm/restart.c b/arch/arm/mach-msm/restart.c
index 7daf77d..18982fa 100644
--- a/arch/arm/mach-msm/restart.c
+++ b/arch/arm/mach-msm/restart.c
@@ -283,12 +283,16 @@ reset:
 #endif /* CONFIG_LGE_CRASH_HANDLER */
 
 	__raw_writel(0, msm_tmr0_base + WDT0_EN);
-	if (!(machine_is_msm8x60_fusion() || machine_is_msm8x60_fusn_ffa())) {
+	if (!(machine_is_apq8064_som() || 
+			machine_is_msm8x60_fusion() || 
+			machine_is_msm8x60_fusn_ffa())) {
 		mb();
 		__raw_writel(0, PSHOLD_CTL_SU); /* Actually reset the chip */
 		mdelay(5000);
 		pr_notice("PS_HOLD didn't work, falling back to watchdog\n");
 	}
+	pr_notice("PS_HOLD didn't work, falling back to watchdog\n");
+	mdelay(100);
 
 	__raw_writel(1, msm_tmr0_base + WDT0_RST);
 	__raw_writel(5*0x31F3, msm_tmr0_base + WDT0_BARK_TIME);
