diff --git a/arch/arm/mach-msm/board-8064-camera.c b/arch/arm/mach-msm/board-8064-camera.c
index 15fc5a1a..bfd6bb7 100644
--- a/arch/arm/mach-msm/board-8064-camera.c
+++ b/arch/arm/mach-msm/board-8064-camera.c
@@ -498,7 +498,7 @@ static struct msm_camera_gpio_conf apq8064_front_cam_gpio_conf = {
 static struct msm_camera_i2c_conf apq8064_back_cam_i2c_conf = {
 	.use_i2c_mux = 1,
 	.mux_dev = &msm8960_device_i2c_mux_gsbi4,
-	.i2c_mux_mode = MODE_L,
+	.i2c_mux_mode = MODE_DUAL,
 	/*.i2c_mux_mode = MODE_R,*/ /* For SOM only gpio 12,13 are under I2C */
 };
 
@@ -531,7 +531,7 @@ static struct msm_camera_i2c_conf apq8064_front_cam_i2c_conf = {
 	.use_i2c_mux = 1,
 	.mux_dev = &msm8960_device_i2c_mux_gsbi4,
 	/*.i2c_mux_mode = MODE_L,*/
-	.i2c_mux_mode = MODE_R, /* For SOM only gpio 12,13 are under I2C */
+	.i2c_mux_mode = MODE_DUAL, /* For SOM only gpio 12,13 are under I2C */
 };
 
 static struct msm_camera_sensor_flash_data flash_imx135 = {
@@ -698,6 +698,60 @@ static struct msm_camera_sensor_info msm_camera_sensor_mt9m114_data = {
 	.sensor_type = YUV_SENSOR,
 };
 
+#ifdef CONFIG_QS_IMX111
+#ifdef CONFIG_QS_IMX111_ACT
+static struct i2c_board_info qs_imx111_actuator_i2c_info = {
+	I2C_BOARD_INFO("qs_imx111_act", 0x18),
+};
+
+static struct msm_actuator_info qs_imx111_actuator_info = {
+	.board_info     = &qs_imx111_actuator_i2c_info,
+	.bus_id         = APQ_8064_GSBI4_QUP_I2C_BUS_ID,
+	.vcm_pwd        = 0,
+	.vcm_enable     = 1,
+};
+#endif
+
+static struct msm_camera_sensor_flash_data flash_qs_imx111 = {
+	.flash_type	= MSM_CAMERA_FLASH_NONE,
+};
+
+/* TBD: csi lane params are missing for this sensor */
+static struct msm_camera_csi_lane_params qs_imx111_csi1_lane_params = {
+    .csi_lane_assign = 0xE4,
+    .csi_lane_mask = 0x3,
+};
+
+static struct msm_camera_csi_lane_params qs_imx111_csi0_lane_params = {
+    .csi_lane_assign = 0x9C,
+    .csi_lane_mask = 0x9,
+};
+
+static struct msm_camera_sensor_platform_info sensor_board_info_qs_imx111 = {
+	.mount_angle	= 90,
+	.gpio_conf = &apq8064_back_cam_gpio_conf,
+	.i2c_conf = &apq8064_back_cam_i2c_conf,
+    .csi_lane_params = &qs_imx111_csi0_lane_params,
+    /*.csi_lane_params = &qs_imx111_csi1_lane_params,*/
+};
+
+static struct msm_camera_sensor_info msm_camera_sensor_qs_imx111_data = {
+	.sensor_name	= "qs_imx111",
+	.pdata	= &msm_camera_csi_device_data[0],
+	.flash_data	= &flash_qs_imx111,
+	/*.strobe_flash_data = &strobe_flash_xenon,*/
+	.sensor_platform_info = &sensor_board_info_qs_imx111,
+	/*.csi_if	= 1,*/
+	.csi_if	= 2,
+/*	.camera_type = BACK_CAMERA_2D,*/
+	.camera_type = BACK_CAMERA_INT_3D,
+#ifdef CONFIG_QS_IMX111_ACT
+	.actuator_info = &qs_imx111_actuator_info
+#endif
+	.sensor_type = BAYER_SENSOR,
+};
+#endif
+
 #ifdef CONFIG_OV5640
 static struct msm_camera_sensor_flash_data flash_ov5640 = {
     .flash_type = MSM_CAMERA_FLASH_NONE
@@ -827,6 +881,12 @@ static struct i2c_board_info apq8064_camera_i2c_boardinfo[] = {
 	.platform_data = &msm_camera_sensor_s5k3l1yx_data,
 	},
 #endif
+#ifdef CONFIG_QS_IMX111
+	{
+	I2C_BOARD_INFO("qs_imx111", 0x34),
+	.platform_data = &msm_camera_sensor_qs_imx111_data,
+	},
+#endif
 };
 
 struct msm_camera_board_info apq8064_camera_board_info = {
