diff --git a/arch/arm/mach-msm/board-8064.c b/arch/arm/mach-msm/board-8064.c
index f1a0c180..d219292 100644
--- a/arch/arm/mach-msm/board-8064.c
+++ b/arch/arm/mach-msm/board-8064.c
@@ -3542,7 +3542,17 @@ static void __init apq8064_cdp_init(void)
 
    if(machine_is_apq8064_som()) {
        int ret = 0;
-       int gpio_line = 86;
+       int gpio_line = 0;
+
+       gpio_line = 47;
+       pr_info("stereo card: hwenable\n");
+       ret = gpio_request(gpio_line, "stereo_hwenable");
+       if (ret)
+           pr_err("%s: Failed to request gpio stereo_hwenable %d\n", __func__, gpio_line);
+       else
+           gpio_direction_output(gpio_line, 1);
+
+       gpio_line = 86;
        pr_info("bringing fpga out of reset\n");
        ret = gpio_request(gpio_line, "fpga_reset_n");
        if (ret)
