diff --git a/kernel/main.c b/kernel/main.c
index 1167f31..3ec5580 100644
--- a/kernel/main.c
+++ b/kernel/main.c
@@ -77,6 +77,7 @@ void kmain(void)
 	target_early_init();
 
 	dprintf(INFO, "welcome to lk\n\n");
+dprintf(INFO, "### Configured for qr-linux ###");
 	
 	// deal with any static constructors
 	dprintf(SPEW, "calling constructors\n");
diff --git a/project/msm8960.mk b/project/msm8960.mk
index 0de2066..0522f1d 100644
--- a/project/msm8960.mk
+++ b/project/msm8960.mk
@@ -11,3 +11,4 @@ DEBUG := 1
 #DEFINES += WITH_DEBUG_DCC=1
 DEFINES += WITH_DEBUG_UART=1
 #DEFINES += WITH_DEBUG_FBCON=1
+CFLAGS += -fno-stack-protector
diff --git a/target/msm8960/atags.c b/target/msm8960/atags.c
index 8ac045c..6d3698e 100755
--- a/target/msm8960/atags.c
+++ b/target/msm8960/atags.c
@@ -246,5 +246,5 @@ void *target_get_scratch_address(void)
 
 unsigned target_get_max_flash_size(void)
 {
-	return (SIZE_768M);
+	return (2*1024*SIZE_1M);
 }
