From da4ee19a32251d3f48f3c09e60a759fb92acddd2 Mon Sep 17 00:00:00 2001
From: Gustavo Solaira <gustavos@codeaurora.org>
Date: Mon, 13 Jul 2015 19:01:53 -0700
Subject: [PATCH 2/3] Explicitly set arm mode if THUMB support is disabled

---
 arch/arm/rules.mk | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/rules.mk b/arch/arm/rules.mk
index d822430..783d24f 100644
--- a/arch/arm/rules.mk
+++ b/arch/arm/rules.mk
@@ -78,6 +78,8 @@ THUMBINTERWORK :=
 ifeq ($(ENABLE_THUMB),true)
 THUMBCFLAGS := -mthumb -D__thumb__
 THUMBINTERWORK := -mthumb-interwork
+else
+CFLAGS += -marm
 endif
 
 INCLUDES += \
-- 
2.4.2

