From b0b4f94da045480732c41534fed60198dc677bc4 Mon Sep 17 00:00:00 2001
From: Kaustubh Gondkar <kgondkar@codeaurora.org>
Date: Thu, 24 Oct 2013 12:45:33 -0700
Subject: [PATCH 3/3] SOM Carrier RevB board serial port rework fix with LVS5
 regulator

---
 target/msm8960/init.c | 18 +++++++++++-------
 1 file changed, 11 insertions(+), 7 deletions(-)

diff --git a/target/msm8960/init.c b/target/msm8960/init.c
index f8dbc3e..0068ca4 100755
--- a/target/msm8960/init.c
+++ b/target/msm8960/init.c
@@ -97,13 +97,7 @@ void target_init(void)
 	unsigned platform_id = board_platform_id();
 
 	dprintf(INFO, "target_init()\n");
-	dprintf(INFO, "Platform ID : (%u)\n", platform_id);
-	dprintf(INFO, "Hardware ID : (%u)\n", board_hardware_id());
-	dprintf(INFO, "Target ID   : (%u)\n", board_target_id());
-	dprintf(INFO, "Baseband    : (%u)\n", board_baseband());
-	dprintf(INFO, "PMIC type   : (%u)\n", board_pmic_type());
-	dprintf(INFO, "PMIC ver    : (%u)\n", board_pmic_ver());
-
+	
 	/* Initialize PMIC driver */
 	pmic.read = (pm8921_read_func) & pa1_ssbi2_read_bytes;
 	pmic.write = (pm8921_write_func) & pa1_ssbi2_write_bytes;
@@ -148,6 +142,16 @@ void target_init(void)
 		dprintf(CRITICAL,"Keyboard is not supported for platform: %d\n",platform_id);
 	};
 
+	/* SOM RevB carrier: Serial port workaround, Enable LVS5 */
+	pm8921_low_voltage_switch_enable(lvs_5); 
+
+	dprintf(INFO, "Platform ID : (%u)\n", platform_id);
+	dprintf(INFO, "Hardware ID : (%u)\n", board_hardware_id());
+	dprintf(INFO, "Target ID   : (%u)\n", board_target_id());
+	dprintf(INFO, "Baseband    : (%u)\n", board_baseband());
+	dprintf(INFO, "PMIC type   : (%u)\n", board_pmic_type());
+	dprintf(INFO, "PMIC ver    : (%u)\n", board_pmic_ver());
+
 	/* Display splash screen if enabled */
 #if DISPLAY_SPLASH_SCREEN
 	display_init();
-- 
1.8.2.1

