From 0eaf87a45163c093368a25941339602ae6a5d89c Mon Sep 17 00:00:00 2001
From: Kaustubh Gondkar <kgondkar@codeaurora.org>
Date: Tue, 5 Nov 2013 14:59:04 -0800
Subject: [PATCH] SOM UART_4wire on GSBI1 as default serial port

---
 target/msm8960/init.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/target/msm8960/init.c b/target/msm8960/init.c
index 0068ca4..e78ed1c 100755
--- a/target/msm8960/init.c
+++ b/target/msm8960/init.c
@@ -316,7 +316,9 @@ void target_uart_init(void)
 		break;
 	case LINUX_MACHTYPE_8064_SOM:
 		/* SOM: UART_2x, GSBI-6, GPIO 14,15 */
-		uart_dm_init(6, 0x16500000, 0x16540000);
+		/*uart_dm_init(6, 0x16500000, 0x16540000);*/
+		/* SOM: UART_2x, GSBI-1, GPIO 18,19 */
+		uart_dm_init(1, 0x12440000, 0x12450000);
 		break;
 	case LINUX_MACHTYPE_8064_MPQ_CDP:
 	case LINUX_MACHTYPE_8064_MPQ_HRD:
-- 
1.8.2.1

