From efe44e830eb454073c039adf37cedbc0a49a75b8 Mon Sep 17 00:00:00 2001
From: Kaustubh Gondkar <kgondkar@codeaurora.org>
Date: Mon, 7 Apr 2014 16:20:53 -0700
Subject: [PATCH] Product line update

---
 Kbuild | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/Kbuild b/Kbuild
index f277128..583d630 100644
--- a/Kbuild
+++ b/Kbuild
@@ -10,7 +10,8 @@ ifeq ($(KERNEL_BUILD),1)
 	# These are provided in Android-based builds
 	# Need to explicitly define for Kernel-based builds
 	MODNAME := wlan
-	WLAN_ROOT := drivers/staging/prima
+	WLAN_ROOT := $(M)
+	obj-m := $(MODNAME).o
 endif
 
 ifeq ($(KERNEL_BUILD),0)
@@ -34,7 +35,6 @@ ifeq ($(KERNEL_BUILD),0)
 
 	#Flag to enable Fast Transition (11r) feature
 	CONFIG_QCOM_VOWIFI_11R := y
-
 endif
 
 # Feature flags which are not (currently) configurable via Kconfig
@@ -516,8 +516,8 @@ CDEFINES :=	-DANI_BUS_TYPE_PLATFORM=1 \
 		-DWLAN_NS_OFFLOAD \
 		-DWLAN_ACTIVEMODE_OFFLOAD_FEATURE \
         	-DWLAN_FEATURE_HOLD_RX_WAKELOCK \
-		-DWLAN_SOFTAP_VSTA_FEATURE \
-                -DWLAN_FEATURE_ROAM_SCAN_OFFLOAD
+		-DWLAN_SOFTAP_VSTA_FEATURE 
+#        -DWLAN_FEATURE_ROAM_SCAN_OFFLOAD
 
 ifneq ($(CONFIG_PRONTO_WLAN),)
 CDEFINES += -DWCN_PRONTO
-- 
1.8.2.1

