From dc5892f1c6cddae2f820cb89468b1d6daa632606 Mon Sep 17 00:00:00 2001
From: Gustavo Solaira <gustavos@codeaurora.org>
Date: Wed, 8 Apr 2015 21:02:55 -0700
Subject: [PATCH 2/3] Fix compilation on OE

---
 Kbuild | 46 +++++++++++++++++++++++-----------------------
 1 file changed, 23 insertions(+), 23 deletions(-)

diff --git a/Kbuild b/Kbuild
index 9906d76..b7b5042 100644
--- a/Kbuild
+++ b/Kbuild
@@ -1,30 +1,30 @@
 ifeq ($(BUILD_ATH6KL_VER_35), 1)
 obj-m += wlan.o
 
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/debug.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/bmi.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/cfg80211.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/cfg80211_btcoex.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/init.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/main.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/txrx.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/wmi.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/wmi_btcoex.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/testmode.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/rttm.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/diag.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/htcoex.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/ath_netlink.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/pm.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/p2p.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/ap.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/debug.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/bmi.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/cfg80211.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/cfg80211_btcoex.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/init.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/main.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/txrx.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/wmi.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/wmi_btcoex.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/testmode.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/rttm.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/diag.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/htcoex.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/ath_netlink.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/pm.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/p2p.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/ap.o
 
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/msm.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/htc.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/hif.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/sdio.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/reg.o
-wlan-y += src/drivers/net/wireless/ath/ath6kl-3.5/regdb.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/msm.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/htc.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/hif.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/sdio.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/reg.o
+wlan-y += drivers/net/wireless/ath/ath6kl-3.5/regdb.o
 endif
 
 ifeq ($(HAVE_CFG80211), 1)
-- 
2.4.2

