From aeb32ad606ceb3efe1c905510d65ad9fe9939410 Mon Sep 17 00:00:00 2001
From: kdlee <kdlee@qti.qualcomm.com>
Date: Tue, 25 Nov 2014 12:35:58 -0800
Subject: [PATCH] use only 1 line for kver

---
 tools/configme | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/configme b/tools/configme
index 7ed9c28..afedc0f 100755
--- a/tools/configme
+++ b/tools/configme
@@ -58,7 +58,7 @@ exit 1
 run_board_config()
 {
     # Can't set these until we've unwound the checkpoint and have meta data.
-    KVER=`cat ./$META_DIR/cfg/kernel-*cache/kver|sed 's/^v//'`
+    KVER=`cat ./$META_DIR/cfg/kernel-*cache/kver|sed 's/^v//'|head -n1`
 
     META=./$META_DIR/meta-series
 
-- 
1.7.11.4

