diff --git a/drivers/net/wireless/ath/ath6kl-3.5/sdio.c b/drivers/net/wireless/ath/ath6kl-3.5/sdio.c
index ba84ed8..0d3cbd6 100755
--- a/drivers/net/wireless/ath/ath6kl-3.5/sdio.c
+++ b/drivers/net/wireless/ath/ath6kl-3.5/sdio.c
@@ -722,7 +722,7 @@ static int ath6kl_sdio_enable_scatter(struct ath6kl *ar)
 {
 	struct ath6kl_sdio *ar_sdio = ath6kl_sdio_priv(ar);
 	struct htc_target *target = ar->htc_target;
-	int ret;
+	int ret = 0;
 	bool virt_scat = false;
 
 	if (ar_sdio->scatter_enabled)
