diff --git a/dpkg-deb/build.c b/dpkg-deb/build.c
index 2fe18ce..ea80ce9 100644
--- a/dpkg-deb/build.c
+++ b/dpkg-deb/build.c
@@ -599,7 +599,7 @@ do_build(const char *const *argv)
       ohshite(_("failed to chdir to `%.255s'"), dir);
     if (chdir(BUILDCONTROLDIR))
       ohshite(_("failed to chdir to `%.255s'"), ".../DEBIAN");
-    execlp(TAR, "tar", "-cf", "-", "--format=gnu", ".", NULL);
+    execlp(TAR, "tar", "-cf", "-", "--format=gnu", "--mtime='TZ=\"UTC\" 1970-01-01'", "--owner=0", "--group=0", ".", NULL);
     ohshite(_("unable to execute %s (%s)"), "tar -cf", TAR);
   }
   close(p1[1]);
@@ -652,7 +652,7 @@ do_build(const char *const *argv)
     m_dup2(p2[1],1); close(p2[0]); close(p2[1]);
     if (chdir(dir))
       ohshite(_("failed to chdir to `%.255s'"), dir);
-    execlp(TAR, "tar", "-cf", "-", "--format=gnu", "--null", "-T", "-", "--no-recursion", NULL);
+    execlp(TAR, "tar", "-cf", "-", "--format=gnu", "--null", "-T", "-", "--no-recursion", "--mtime='TZ=\"UTC\" 1970-01-01'", "--owner=0", "--group=0", NULL);
     ohshite(_("unable to execute %s (%s)"), "tar -cf", TAR);
   }
   close(p1[0]);
diff --git a/lib/dpkg/ar.c b/lib/dpkg/ar.c
index 808ef58..66ac4a7 100644
--- a/lib/dpkg/ar.c
+++ b/lib/dpkg/ar.c
@@ -100,7 +100,7 @@ dpkg_ar_member_put_header(const char *ar_name, int ar_fd,
 		ohshit(_("ar member size %jd too large"), size);
 
 	n = sprintf(header, "%-16s%-12lu0     0     100644  %-10jd`\n",
-	            name, time(NULL), (intmax_t)size);
+	            name, 0, (intmax_t)size);
 	if (n != sizeof(struct ar_hdr))
 		ohshit(_("generated corrupt ar header for '%s'"), ar_name);
 
