diff --git a/drivers/video/msm/mdp4_dtv.c b/drivers/video/msm/mdp4_dtv.c
index ba810a9..be16d3b 100644
--- a/drivers/video/msm/mdp4_dtv.c
+++ b/drivers/video/msm/mdp4_dtv.c
@@ -173,6 +173,11 @@ static int dtv_on(struct platform_device *pdev)
 	mfd = platform_get_drvdata(pdev);
 	panel_pixclock_freq = mfd->fbi->var.pixclock;
 
+   if ( mfd->fbi->var.pixclock > 148500000 )
+   {
+      mfd->fbi->var.pixclock = 148500000;
+   }
+
 	if (panel_pixclock_freq > 58000000)
 		/* pm_qos_rate should be in Khz */
 		pm_qos_rate = panel_pixclock_freq / 1000 ;
diff --git a/drivers/video/msm/msm_fb.c b/drivers/video/msm/msm_fb.c
index 6d2a537..3b1eeb9 100644
--- a/drivers/video/msm/msm_fb.c
+++ b/drivers/video/msm/msm_fb.c
@@ -1327,7 +1327,8 @@ static int msm_fb_register(struct msm_fb_data_type *mfd)
 		bpp = 3;
 		break;
 
-	case MDP_ARGB_8888:
+	case MDP_RGBA_8888:
+      pr_err( "MODE MDP_RGBA_8888" );
 		fix->type = FB_TYPE_PACKED_PIXELS;
 		fix->xpanstep = 1;
 		fix->ypanstep = 1;
@@ -1335,6 +1336,7 @@ static int msm_fb_register(struct msm_fb_data_type *mfd)
 		var->blue.offset = 0;
 		var->green.offset = 8;
 		var->red.offset = 16;
+
 		var->blue.length = 8;
 		var->green.length = 8;
 		var->red.length = 8;
@@ -1346,7 +1348,8 @@ static int msm_fb_register(struct msm_fb_data_type *mfd)
 		bpp = 4;
 		break;
 
-	case MDP_RGBA_8888:
+	case MDP_ARGB_8888:
+      pr_err( "MODE MDP_ARGB_8888" );
 		fix->type = FB_TYPE_PACKED_PIXELS;
 		fix->xpanstep = 1;
 		fix->ypanstep = 1;
@@ -1354,6 +1357,7 @@ static int msm_fb_register(struct msm_fb_data_type *mfd)
 		var->blue.offset = 8;
 		var->green.offset = 16;
 		var->red.offset = 24;
+
 		var->blue.length = 8;
 		var->green.length = 8;
 		var->red.length = 8;
diff --git a/drivers/video/msm/mdp4_overlay.c b/drivers/video/msm/mdp4_overlay.c
index 7f1cd0a..cda67c2 100644
--- a/drivers/video/msm/mdp4_overlay.c
+++ b/drivers/video/msm/mdp4_overlay.c
@@ -1239,7 +1239,7 @@ int mdp4_overlay_format2pipe(struct mdp4_overlay_pipe *pip
                pipe->element0 = C3_ALPHA;      /* alpha */
                pipe->bpp = 4;          /* 4 bpp */
                break;
-       case MDP_RGBA_8888:
+       case MDP_BGRA_8888:
                pipe->frame_format = MDP4_FRAME_FORMAT_LINEAR;
                pipe->fetch_plane = OVERLAY_PLANE_INTERLEAVED;
                pipe->a_bit = 3;        /* alpha, 4 bits */
@@ -1273,7 +1273,7 @@ int mdp4_overlay_format2pipe(struct mdp4_overlay_pipe *pip
                pipe->element0 = C2_R_Cr;       /* R */
                pipe->bpp = 4;          /* 4 bpp */
                break;
-       case MDP_BGRA_8888:
+       case MDP_RGBA_8888:
                pipe->frame_format = MDP4_FRAME_FORMAT_LINEAR;
                pipe->fetch_plane = OVERLAY_PLANE_INTERLEAVED;
                pipe->a_bit = 3;        /* alpha, 4 bits */

