diff --git a/drivers/bluetooth/Kconfig b/drivers/bluetooth/Kconfig
index 5ccf142..8e81572 100644
--- a/drivers/bluetooth/Kconfig
+++ b/drivers/bluetooth/Kconfig
@@ -229,4 +229,21 @@ config BT_WILINK
 
 	  Say Y here to compile support for Texas Instrument's WiLink7 driver
 	  into the kernel or say M to compile it as module.
+
+config MSM_BT_POWER
+   tristate "MSM Bluetooth Power Control"
+   depends on ARCH_MSM && RFKILL
+   default m
+   help
+     Provides a parameter to switch on/off power from PMIC
+     to Bluetooth device.
+
+config BT_MSM_SLEEP
+   tristate "MSM Bluesleep driver"
+   depends on BT && SERIAL_MSM_HS
+   default n
+   help
+     Bluetooth MSM bluesleep driver.
+     This driver provides support for BTS sleep.
+
 endmenu
diff --git a/drivers/bluetooth/Makefile b/drivers/bluetooth/Makefile
index f4460f4..7dfa786d 100644
--- a/drivers/bluetooth/Makefile
+++ b/drivers/bluetooth/Makefile
@@ -29,3 +29,6 @@ hci_uart-$(CONFIG_BT_HCIUART_BCSP)	+= hci_bcsp.o
 hci_uart-$(CONFIG_BT_HCIUART_LL)	+= hci_ll.o
 hci_uart-$(CONFIG_BT_HCIUART_ATH3K)	+= hci_ath.o
 hci_uart-objs				:= $(hci_uart-y)
+msm_bt_sleep-objs                       := bluesleep.o
+obj-$(CONFIG_MSM_BT_POWER)    += bluetooth-power.o
+
