From 5da7307754bbc8d30730bff89f8cdd60b5432cfd Mon Sep 17 00:00:00 2001
From: Kevin Lee <kdlee@codeaurora.org>
Date: Fri, 18 Apr 2014 17:19:10 -0700
Subject: [PATCH] Fix the ldd LTP test

---
 testcases/commands/ade/ldd/ldd01 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/testcases/commands/ade/ldd/ldd01 b/testcases/commands/ade/ldd/ldd01
index 8c74389..70ef766 100755
--- a/testcases/commands/ade/ldd/ldd01
+++ b/testcases/commands/ade/ldd/ldd01
@@ -63,7 +63,7 @@ do_setup()
 	$CC $CFLAGS -shared -o $TCtmp/lddfile3.obj.so $TCdat/lddfile3.obj
 	$CC $CFLAGS -shared -o $TCtmp/lddfile4.obj.so $TCdat/lddfile4.obj
 	$CC $CFLAGS -shared -o $TCtmp/lddfile5.obj.so $TCdat/lddfile5.obj
-	$CC $CFLAGS -O -o $TCtmp/a.out $TCtmp/lddfile*.obj.so $TCdat/main.obj
+	$CC $CFLAGS -O -o $TCtmp/a.out $TCdat/main.obj $TCtmp/lddfile*.obj.so
 }
 
 do_test()
-- 
1.7.11.4

