From 8698e9db9cb1d11631ad68cad7bb102c20d51183 Mon Sep 17 00:00:00 2001
From: Kevin Lee <kdlee@codeaurora.org>
Date: Thu, 1 May 2014 13:17:18 -0700
Subject: [PATCH] Fix LTP profil01 test by turning off compiler optimization

---
 testcases/kernel/syscalls/profil/Makefile | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/testcases/kernel/syscalls/profil/Makefile b/testcases/kernel/syscalls/profil/Makefile
index bd617d8..7faee3a 100644
--- a/testcases/kernel/syscalls/profil/Makefile
+++ b/testcases/kernel/syscalls/profil/Makefile
@@ -18,6 +18,12 @@
 
 top_srcdir		?= ../../../..
 
+
+
+
 include $(top_srcdir)/include/mk/testcases.mk
 
+CFLAGS := -g
+
 include $(top_srcdir)/include/mk/generic_leaf_target.mk
+
-- 
1.7.11.4

