From 9622d4db231c94a5545556135a5952902219595c Mon Sep 17 00:00:00 2001
From: Kevin Lee <kdlee@codeaurora.org>
Date: Wed, 18 Jun 2014 13:32:39 -0700
Subject: [PATCH] Fix LTP mount03 test

---
 testcases/kernel/syscalls/mount/mount03.c | 24 +++++++++++++++++-------
 1 file changed, 17 insertions(+), 7 deletions(-)

diff --git a/testcases/kernel/syscalls/mount/mount03.c b/testcases/kernel/syscalls/mount/mount03.c
index 04570b9..d057299 100644
--- a/testcases/kernel/syscalls/mount/mount03.c
+++ b/testcases/kernel/syscalls/mount/mount03.c
@@ -226,7 +226,7 @@ int main(int ac, char **av)
 
 int test_rwflag(int i, int cnt)
 {
-	int fd, pid, status;
+	int fd, pid, status, ret;
 	char nobody_uid[] = "nobody";
 	struct passwd *ltpuser;
 
@@ -276,7 +276,10 @@ int test_rwflag(int i, int cnt)
 			tst_resm(TWARN | TERRNO, "opening %s failed", file);
 		} else {
 			close(fd);
-			execlp(file, basename(file), NULL);
+			ret=execlp(file, basename(file), NULL);
+                        if ((ret == -1) && (errno == EACCES))
+				return 0;
+			tst_resm(TWARN | TERRNO, "testing MS_NOEXEC %s failed", file);
 		}
 		return 1;
 	case 3:
@@ -431,6 +434,8 @@ int setup_uid()
 /* setup() - performs all ONE TIME setup for this test */
 void setup()
 {
+        int fd;
+        char path[PATH_MAX];
 	char *test_home;	/* variable to hold TESTHOME env */
 	struct stat setuid_test_stat;
 
@@ -442,12 +447,12 @@ void setup()
 		tst_brkm(TBROK, NULL, "Test must be run as root");
 	}
 
-	/* Test home directory */
-	test_home = get_current_dir_name();
-
 	/* make a temp directory */
 	tst_tmpdir();
 
+	/* Test home directory */
+	test_home = get_current_dir_name();
+
 	/* Unique mount point */
 	(void)sprintf(mntpoint, "mnt_%d", getpid());
 
@@ -463,7 +468,11 @@ void setup()
 		tst_brkm(TBROK, cleanup, "chmod(%s, %#o) failed",
 			 Path_name, DIR_MODE);
 	}
-	snprintf(file, PATH_MAX, "%ssetuid_test", Path_name);
+	snprintf(file, PATH_MAX, "%s/setuid_test", Path_name);
+        fd = open(file, O_CREAT | O_TRUNC);
+  	if (fd == -1)
+  		tst_brkm(TBROK, cleanup, "open file failed");
+  	close(fd);
 	if (stat(file, &setuid_test_stat) < 0) {
 		tst_brkm(TBROK, cleanup, "stat for setuid_test failed");
 	} else {
@@ -482,7 +491,8 @@ void setup()
 	/*
 	 * under temporary directory
 	 */
-	snprintf(Path_name, PATH_MAX, "%s/%s/", Path_name, mntpoint);
+	strcpy(path, Path_name);
+	snprintf(Path_name, PATH_MAX, "%s/%s/", path, mntpoint);
 
 	strcpy(testhome_path, test_home);
 	strcat(testhome_path, "/setuid_test");
-- 
1.7.11.4

